---
title: "Figure 3 and S1/2/4"
author: "Salvatore Milite"
date: "12/21/2021"
output: html_document
---

## Figure 3 ans S4

```{r}
### FUNCTION TO PLOT CCF HISTOGRAMS ###

plot_hist <- function(data) {
  
  library(ggrepel)
  
  clusters <- data %>% group_by(cluster) %>%  summarise(mean_ccf = mean(ccf.y %>% as.numeric)) %>% arrange(-mean_ccf)
  
  clusters$new_clusters <-  seq(1:nrow(clusters)) %>%  paste
  
  data_1 <- inner_join(data,clusters) %>% select(ccf.y, new_clusters, name, isdriver, Hugo_Symbol, clonality)
  
  data_1$pipeline <-  "PCWAG"
  
  clusters_2 <- data %>% group_by(cluster_id) %>%  summarise(mean_ccf = cellular_prevalence %>% unique()) %>% arrange(-mean_ccf)
  clusters_2$new_clusters <-  seq(1:nrow(clusters_2)) %>%  paste
  
  data_2 <- inner_join(data,clusters_2)%>% select(ccf.y, new_clusters, name, isdriver, Hugo_Symbol, clonality)
  
  data_2$pipeline <-  "Pyclone"
  
  data_plot <- rbind(data_1,data_2)
  
  ggplot(data = data_plot, aes(x = ccf.y %>% as.numeric, fill = new_clusters %>%  paste)) +
    geom_histogram(bins = 100, position = "stack", alpha = 0.6) +
    geom_text_repel(data = data_plot %>%  filter(!is.na(isdriver),  clonality == "subclonal"),
                    aes(label = Hugo_Symbol, y = Inf, x = ccf.y %>%  as.numeric), colour = "black",
                    label.size = 0.01, alpha = 0.85) +
    geom_vline(data = data_plot %>%  filter(!is.na(isdriver), clonality == "subclonal"),
               aes(xintercept = ccf.y %>% as.numeric, colour = new_clusters %>%  paste), lty = 2, show.legend = FALSE) +
    scale_fill_brewer("cluster",palette = "Dark2", aesthetics = c("colour", "fill")) +
    facet_wrap(.~pipeline, scales = "free_y") + ggtitle(data_plot$name %>% unique()) +
    theme_bw() + theme(legend.position = "None", plot.title = element_text(hjust = 0.5),text = element_text(size=20) ) +
    xlim(c(0,2)) +
    xlab("CCF") +
    ylab("Counts") 
}
```

This two dataset can be obtained respectively by running the "PCAWG data" and "Pyclone reclustering"

```{r, eval = FALSE}
library(tidyverse)

all_pyclone <- readRDS("./pyclone_refits/pyclone_BB_all.rds")
all_PCWAG <- data.table::fread("../maf_subclone_ccf.csv.gz", data.table = F)
```


```{r}
# This file is publicly downloadable at this link https://dcc.icgc.org/releases/PCAWG/evolution_and_heterogeneity
annots <- data.table::fread("../icgc_sample_annotations_summary_table.txt", data.table=F) %>% rename(sample_id = tumour_aliquot_id) %>%
  mutate(name = paste0(histology_abbreviation, "\n",icgc_sample_id)) %>%
  select(sample_id, name)

# We have a subset of manually choosen 41 samples
load("../samples_letter.rda", verbose = T)
```

```{r}
all_pyclone_letter <- all_pyclone %>%  do.call(rbind,.) %>% filter(sample_id %in% nms_letter)

p3_1_df <- all_pyclone_letter %>%  dplyr::inner_join(annots, .)

p3_1_df <- dplyr::inner_join(all_PCWAG %>% mutate(mutation_id = paste(paste0("chr",Chromosome), Start_position, Start_position, round(ccf,3), gene, is_driver, sep = ":")),
                             p3_1_df , by ="mutation_id")

# Also interesting to look just at SNVs that usually have a higher mapping quality
#p3_1_df <-  p3_1_df %>%  filter(Variant_Type== "SNP")


# We order the clusters based on ccf
new_clusters <-  p3_1_df %>% group_by(name, cluster_id) %>% summarize(ccf_cluster = unique(cellular_prevalence) ) %>% arrange(-ccf_cluster) %>% ungroup() %>%
  group_by(name) %>% 
  mutate(new_cluster_ids = order(ccf_cluster, decreasing = T))

p3_supp_df <- p3_1_df %>% inner_join(., new_clusters)
```
With this we generate Supplementary Figure S4 (41 cases clustered by Pyclone-VI)

```{r}
library(ggrepel)

p3_supp <- ggplot(data = p3_supp_df, aes(x = ccf.y %>% as.numeric, fill = new_cluster_ids %>%  paste)) +
  geom_histogram(bins = 100, position = "stack", alpha = 0.6) +
 geom_text_repel(data = p3_supp_df %>%  filter(!is.na(isdriver),  clonality == "subclonal"),
                  aes(label = Hugo_Symbol, y = Inf, x = ccf.y %>%  as.numeric), colour = "black",
                  size = 2, alpha = 0.85) +
  geom_vline(data = p3_supp_df %>%  filter(!is.na(isdriver), clonality == "subclonal"),
             aes(xintercept = ccf.y %>% as.numeric, colour = new_cluster_ids %>%  paste), lty = 2, show.legend = FALSE) +
  scale_fill_brewer("cluster",palette = "Dark2", aesthetics = c("colour", "fill")) +
  facet_wrap(.~name, scales = "free_y") + 
  theme_bw() + theme(legend.position = "None") +
  xlim(c(0,2)) +
  xlab("CCF") +
  ylab("Counts") + ggtitle("Pyclone-vi clustering")


p3_supp %>% ggsave(.,filename = "figure_S4.png", device = "png", units = "px", width = 3200, height = 2500)

```

For figure 6 we first generate panels a-g

```{r}
# We assemble the 6 plots of figure a manually

samples_to_plot <- c("Panc-AdenoCA\nSA410399", "Panc-AdenoCA\nSA410978", "Panc-AdenoCA\nSA520252", "Panc-AdenoCA\nSA520283",
                     "Biliary-AdenoCA\nSA543550", "Liver-HCC\nSA501649")

p_3_1_1 <- plot_hist(p3_1_df %>% filter(name == samples_to_plot[1]))

p_3_1_2 <- plot_hist(p3_1_df %>% filter(name == samples_to_plot[2]))

p_3_1_3 <- plot_hist(p3_1_df %>% filter(name == samples_to_plot[3]))

p_3_1_4 <- plot_hist(p3_1_df %>% filter(name == samples_to_plot[4]))

p_3_1_5 <- plot_hist(p3_1_df %>% filter(name == samples_to_plot[5]))

p_3_1_6 <- plot_hist(p3_1_df %>% filter(name == samples_to_plot[6]))

library(patchwork)

p3_1 <- (p_3_1_1 | p_3_1_2 |p_3_1_3 ) / (p_3_1_4 | p_3_1_5 | p_3_1_6)

```

Then we plot panels 

```{r}

p3_2_df1 <-  all_PCWAG %>% filter(isdriver == T) %>%  group_by(clonality) %>%  summarize(N = n()) %>% filter(clonality %in% c("clonal", "subclonal"))

p3_2_df2 <- all_pyclone %>%  do.call(rbind,.) 

p3_2_df2 <- dplyr::inner_join(all_PCWAG %>% select(-clonality, - gene )%>%  mutate(bind_id = paste(paste0("chr",Chromosome), Start_position, Start_position, round(ccf,3),gene, is_driver, sep = ":")),
                              p3_2_df2), by ="mutation_id")

# Here we somehow copy the assignment criteria of the Dentro paper, a pyclone cluster is clonal if it is the max(cellular prevalence) cluster
# or it has ccf > 90%  

p3_2_df2 <- p3_2_df2 %>% group_by(sample_id) %>% mutate(clonal_ccfs = max(cellular_prevalence))

p3_2_df2 <-  p3_2_df2 %>% mutate(clonality = if_else(cellular_prevalence > 0.9, "clonal", "subclonal")) %>% mutate(clonality = if_else(cellular_prevalence == clonal_ccfs,
                                                                                                                                       "clonal", clonality))
p3_2_df2_aggr <- p3_2_df2 %>% filter(is_driver != "NA") %>%  group_by(clonality) %>%  summarize(N = n()) %>% filter(clonality %in% c("clonal", "subclonal"))

p3_2_df2_aggr$pipeline <- "PyClone"
p3_2_df1$pipeline <-  "PCWAG"

p3_2 <- ggplot(data =  rbind(p3_2_df2_aggr, p3_2_df1)%>% filter(clonality == "subclonal"),
               mapping = aes(x = pipeline , y = N)) +
  geom_col(colour = "black", size = 0.3, fill = "gainsboro", alpha = 0.8) +
  theme_bw() + theme(legend.position = "None", text = element_text(size=20)) +
  xlab("") + ylab("#drivers") + 
  ggtitle("Subclonal drivers (SNVs/indels) in 1340 samples")

p3_3_df1 <- p3_2_df2 %>%  filter(sample_id %in% nms_letter, is_driver != "NA", mut_type == "SNV" ) %>%  group_by(clonality) %>%  summarize(N = n()) %>% filter(clonality %in% c("clonal", "subclonal"))
p3_3_df2 <- all_PCWAG %>% filter(isdriver == T,mut_type == "SNV" ) %>%  filter(Tumor_Sample_Barcode %in% nms_letter) %>%  group_by(clonality) %>%  summarize(N = n()) %>% filter(clonality %in% c("clonal", "subclonal"))

p3_3_df1$pipeline <- "PyClone"
p3_3_df2$pipeline <-  "PCWAG"

p3_3 <- ggplot(data =  rbind(p3_3_df1, p3_3_df2) %>% filter(clonality == "subclonal"),
               mapping = aes(x = pipeline , y = N )) +
  geom_col(colour = "black", size = 0.3, alpha = 0.8, fill = "gainsboro") +
  theme_bw() + theme(legend.position = "None", text = element_text(size=20)) +
  xlab("") + ylab("#drivers") + 
  ggtitle("Subclonal drivers (SNVs/indels) in 41 selected samples")

```

```{r}
p3_4_1 <-  all_pyclone %>%  do.call(rbind,.) 

p3_4_1 <-  p3_4_1 %>% group_by(sample_id) %>% mutate(mut_tot = n()) 


### HERE WE REMOVE THE PYCLONE CLUSTERS WITH MIXING PROPORTION LESS THAN 0.01 AS IN THE DENTRO ANALYSIS ###
p3_4_1 <-  p3_4_1 %>% group_by(sample_id, cluster_id) %>% mutate(mixing_prop = n() / mut_tot) %>%  mutate(to_remove = if_else(mixing_prop < 0.01 , T, F)) 

p3_4_1_aggr <-  p3_4_1 %>% filter(!to_remove) %>% group_by(sample_id) %>% summarize(N = length(unique(cluster_id)))

p3_4_2_aggr <-  all_PCWAG %>% group_by(Tumor_Sample_Barcode) %>% summarize(N = length(unique(cluster)))

p3_4_2_aggr <-  p3_4_2_aggr %>% rename(sample_id = Tumor_Sample_Barcode)

p3_4_2_aggr$pipeline <- "PCWAG"
p3_4_1_aggr$pipeline <- "PyClone"


plot_df <-  rbind(p3_4_2_aggr, p3_4_1_aggr) %>%  mutate(N = N -1) %>% 
  mutate(N = if_else(N > 3, ">3", N %>%  paste))

plot_df$N <- factor(plot_df$N, levels = c("0","1", "2", "3", ">3"))
p3_4 <- ggplot(data =  plot_df, mapping = aes(x = N, fill = N %>%  paste())) +
  geom_bar(colour = "white") +
  scale_fill_brewer("Clonality", palette = "Blues") +
  theme_bw() + theme(legend.position = "None", text = element_text(size=20)) +
  facet_wrap(.~pipeline) +
  xlab("") + ylab("#subclones") + 
  ggtitle("Number of subclones in 1340 samples")
```

## Figure S1


## Figure S2


