---
title: "Overdispersion in germline SNPs"
output: 
  rmdformats::downcute:
    self_contained: true
    thumbnails: true
    lightbox: true
    gallery: false
    highlight: tango


---

## Overdispersion from germline SNPs

the germline analysis the function `generate_report_overdispersion` provides a complete report for overdispersion, with the 6 plots included in the Supplementary Data and binomial and beta-binomial MLE paramters. The input requires 5 objects that can be obtained from the PCWAG data release:

* *snps*: germinal SNPs, they can be obtained from the normal and tumour BAMS (restricted access), the input table should have the 5 following field "chr", "ref", "alt", "AD", "DP". Where "ref" is the reference allele, “alt“ the alternative and "DP" and "AD" can be obtained from the VCF.
*  *cna*: [PCWAG consensus CN files](https://dcc.icgc.org/api/v1/download?fn=/PCAWG/consensus_cnv/consensus.20170119.somatic.cna.annotated.tar.gz)
*  *meta*: a [table](https://dcc.icgc.org/api/v1/download?fn=/PCAWG/consensus_cnv/consensus.20170217.purity.ploidy.txt.gz) with at least a coloumn named `purity` with the actual purity values for the sample 
*  *snvs*: maf_subclone_ccf.csv can be obtained as described in the `Process PCAWG data` article
*  *name*: the sample name (just used to generate plot labels)

If you don't already have the *cna* and *meta* files, you can automatically download them by running this chunk

**The workingdir of this vignette is letter_Dentro_et_al_2021/germline_overdispersion_analysis**


```{r, include = FALSE}
knitr::opts_chunk$set(collapse = TRUE)
```


```{r, eval = FALSE}
options(timeout=100)
```


```{r, eval = FALSE}
download.file("https://dcc.icgc.org/api/v1/download?fn=/PCAWG/consensus_cnv/consensus.20170119.somatic.cna.annotated.tar.gz", destfile = "cnvs.tar.gz")
download.file("https://dcc.icgc.org/api/v1/download?fn=/PCAWG/consensus_cnv/consensus.20170217.purity.ploidy.txt.gz", destfile = "meta.tar.gz")

```

```{r, eval = FALSE}
require(tidyverse)

meta <-  readr::read_tsv("consensus.20170217.purity.ploidy.txt.gz") %>% split(., .$samplename)
untar("consensus.20170119.somatic.cna.annotated.tar.gz", exdir = "cnas")
cnas <-  lapply(list.files(path = "cnas", full.names = T), function(x) readr::read_tsv(x, show_col_types = FALSE))

cna_nms <- list.files(path = "cnas") %>%  substr(1,36)
names(cnas) <-  cna_nms
```

We also provide the list of AD fields for germline SNPs of this 41 samples. To download them

```{r, eval = FALSE}

download.file("https://www.dropbox.com/s/ljyv82l9pkj5yqf/letter_SNPs_anonim.rds?dl=1", destfile = "letter_SNPs.rds")

snps <- readRDS("letter_SNPs.rds")

```


Once you have all the data you can simply run this chunk of code to produce the report as in the `germline_supp_figure.pdf`.
In this vignette we will also run all of the plots separately on a released anonymized version of the SNPs data for a single sample to explain how they are made.

```{r, eval = FALSE}

require(ggrepel)

source("germline_analysis.R")

res <- generate_report_overdispersion(snps, snvs, cna, name, meta)

# overdispersion report
res$plts

# information criteria and lk
res$scores

# Fitted objects more information on the fields here (https://github.com/caravagnalab/BMix)
res$BBMIX
res$BMIX

```

To run it for all the samples (here we use our in house package for palatalization):

```{r, eval = FALSE}
require(easypar)

inputs <- mapply(snps, snvs, cnas, names, meta, function(x1,x2,x3,x4,x5) list(snps = x1, snvs = x2, cna = x3,
                                                                              name = x4, meta = x5), SIMPLIFY = FALSE)

easypar::run(generate_report_overdispersion,
             inputs, parallel = T, 
             filter_errors = TRUE, 
             cores.ratio = 0.5 )


```

## Tarone Z-test

We first have defined two helper function to perform the [Tarone Z-test](https://academic.oup.com/biomet/article-abstract/66/3/585/232596?redirectedFrom=fulltext) . The first one performs the standard Tarone test as defined in the paper. The other one provides a bootstapped version of the test and allows the computation of confidence interval, this second function is conceptually described in the following [blog post](https://rpubs.com/cakapourani/beta-binomial)

```{r}
Tarone.test <- function(N, M) {
  #Check validity of inputs
  if (!(all(N == as.integer(N)))) {
    stop("Error: Number of trials should be integers")

  }
  if (min(N) < 1) {
    stop("Error: Number of trials should be positive")

  }
  if (!(all(M == as.integer(M)))) {
    stop("Error: Count values should be integers")

  }
  if (min(M) < 0) {
    stop("Error: Count values cannot be negative")

  }
  if (any(M > N)) {
    stop("Error: Observed count value exceeds number of trials")

  }
  #Set description of test and data
  method      <- "Tarone's Z test"

  data.name   <- paste0(deparse(substitute(M)),
                        " successes from ",
                        deparse(substitute(N)),
                        " trials")

  #Set null and alternative hypotheses
  null.value  <- 0

  attr(null.value, "names") <- "dispersion parameter"

  alternative <- "greater"

  #Calculate test statistics
  estimate    <- sum(M) / sum(N)

  attr(estimate, "names") <- "proportion parameter"

  S           <- ifelse(estimate == 1, sum(N),
                        sum((M - N * estimate) ^ 2 / (estimate * (1 - estimate))))

  statistic   <- (S - sum(N)) / sqrt(2 * sum(N * (N - 1)))

  attr(statistic, "names") <- "z"

  #Calculate p-value
  p.value     <- 2 * pnorm(-abs(statistic), 0, 1)

  attr(p.value, "names") <- NULL

  #Create htest object
  TEST        <- list(
    method = method,
    data.name = data.name,
    null.value = null.value,
    alternative = alternative,
    estimate = estimate,
    statistic = statistic,
    p.value = p.value
  )

  class(TEST) <- "htest"

  TEST

}

```

```{r}
# TARONE bootstrap test

Tarone_Boostrap <-
  function(x,
           num_samples = 1000)  {
    
    library(tidyverse)


    N = x$DP %>% as.numeric()
    M = x$NV %>% as.numeric()
    z_scores = c()

    for (i in 1:num_samples) {
      index = sample(1:length(N), size = length(N), replace = TRUE)
      n = N[index]
      m = M[index]

      p = sum(m) / sum(n)
      S = ((m - n * p) ^ 2)/(p * (1 - p))
      S = sum(S)
      Z = (S - sum(n)) / sqrt(2 * sum(n * (n - 1)))
      z_scores = c(z_scores, Z)
    }

    log_p_value = dnorm(z_scores, log = TRUE) %>% sum()

    null_hyp = rnorm(num_samples)
    y = data.frame(value = z_scores, data = "Z-score")
    w = data.frame(value = null_hyp, data = "Binomial (sim.)")
    data = rbind(y, w)

    # Create histogram plot
    p_label = format(log_p_value %>% exp(), digits = 3)
    p_label = ifelse(p_label == "0", "p<1e-12", paste("p = ", p_label))

    plot_tarone <-
      ggplot(data, aes(x = value, fill = data)) +
      geom_histogram(
        aes(y = ..density..),
        alpha = 0.6,
        position = "dodge",
        bins = 50
      ) +
      geom_vline(xintercept = 0,
                 linetype = "dashed",
                 size = 0.7) +
      stat_function(
        fun = dnorm,
        colour = "cornflowerblue",
        size = 1,
        args = list(mean = 0, sd = 1)
      ) +
      labs(
        title = paste0("Tarone's Z score; ", p_label),
        x = "log10(Z) (overdispersion)",
        y = "Density"
      ) +
      scale_fill_manual(values = c('Binomial (sim.)' = "#377EB8", "Z-score" = "#E41A1C")) +
      theme_linedraw(base_size = 9) +
      guides(fill = guide_legend("")) +
      theme(legend.position = "bottom")

    plot_tarone

  }

```

## Reproducing supplementary dataset 2

To illustrate how the figure is made, we released a sample dataset for a single patient and we will explain how the panels are made.

We start by plotting the density and the histograms of the fitted distribution. To fit the binomial and beta-binomial here we used the package ![Bmix](https://github.com/caravagnalab/BMix) which for one components performs just a simple numerical optimization of the Negative Log-Likelihood (NLL). Another problem is the presence of homozygous SNPs in around VAF=1 and technical artifact around VAF=0. In order to remove themwe filter SNPs at a frequency < 0.12 and > 0.88, this provides a decent result.

```{r}
require(dplyr)
require(ggplot2)

set.seed(3)

example <- readRDS("germline_SNPs_example.rds")

snps <- example$snps

name <-  "50e3a421-9741-4cd8-911b-0f47ca518b6a"

colnames(snps) <-  c("mut_id", "AD", "DP")

  # Separate the AD filed to obtain DR and DV
snps <-
  snps %>% tidyr::separate(col = "AD",
                           into = c("DR", "DV", "DO"),
                           sep = ",") 

# Filter multiallelic SNPs
snps_filt <- snps %>%  filter(DO == 0)

snps_filt <-
  snps_filt %>% mutate(DV = as.numeric(DV), DP  = as.numeric(DP)) %>% 
  mutate(VAF = DV / DP)

hist(snps_filt$DV / snps_filt$DP)




# Filter hetero SNPs
L_FILT <- 0.12
H_FILT <-  0.88


snps_het <-  snps_filt %>%  filter(VAF > L_FILT, VAF < H_FILT)

hist(snps_het$DV / snps_het$DP)
```

We then performed the Tarone test and the binomial/beta-binomial inference with Bmix.

```{r}

T_test <-  Tarone.test(snps_het$DP
                       ,snps_het$DV )

bmix_data <- data.frame(snps_het$DV , snps_het$DP )

colnames(bmix_data) <- c("DV", "DP")

# For efficiency reasons we subsample the data
if(nrow(bmix_data) > 50000) {

  bmix_data_s <- bmix_data %>% sample_n(50000)

} else {
  bmix_data_s <-  bmix_data
}

# Standard BMix fit, with EM
BMIX_fit_BB <- BMix::bmixfit(K.Binomials = 0, K.BetaBinomials = 1, 
                             samples = 1, score = "NLL",data = bmix_data_s)

BMIX_fit_B1 <- BMix::bmixfit(data = bmix_data_s, K.Binomials = 1, K.BetaBinomials = 0,
                             samples = 1, score = "NLL")


x <-  seq(0,100,by = 1)

# plot the binomial over median coverage
COV <- median(snps_filt$DP) %>%  round

bin1 <-  dbinom(x, size = COV, prob = BMIX_fit_B1$B.params[1]) %>%  data.frame(density = .) %>%
  mutate(weight = 1, type = "bin1", x = x)


bbin <- VGAM::dbetabinom(x, size = COV, prob = BMIX_fit_BB$BB.params[1,1], rho = BMIX_fit_BB$BB.params[2,1]) %>%  data.frame(density = .)%>%
  mutate(weight = 1, type = "bbin", x = x)

density <- rbind(bin1, bbin)


dens_plot <-
  ggplot(density , aes(x = x, y = density * weight, color = type)) + geom_line() + theme_bw() + xlim(0, 100) + xlab("") +
  scale_color_brewer("model", palette = "Set1") + ylab("density") + ggtitle(paste0("bbin vs bin fits (rho= ",round(BMIX_fit_BB$BB.params[2],3),")"))

dens_plot

```

We then compared some commonly used model selection scores among the two fits. The NLL alone usually favors more complex model, the BIC and AIC penalize complexity, the BIC increases its penalty with the number of mutations.

```{r}
BB_S <- data.frame(model = "bbin", NLL2 = BMIX_fit_BB$NLL * 2, AIC =2 * 2 + BMIX_fit_BB$NLL * 2,  BIC = BMIX_fit_BB$BIC)

B1_S <- data.frame(model = "bin1", NLL2 = BMIX_fit_B1$NLL * 2, AIC =2 * 1  +BMIX_fit_B1$NLL * 2,  BIC = BMIX_fit_B1$BIC)

score_df <-  rbind(BB_S, B1_S)

score_df_l <-  reshape2::melt(score_df, id.vars = c("model"))

score_plot <- ggplot(score_df_l, aes(x = variable %>%  as.factor(), y = value %>%  as.numeric(), fill = model %>%  as.factor())) +
  geom_bar(stat="identity", position=position_dodge()) +
  theme_bw() + xlab("") +
  scale_fill_brewer("model", palette = "Set1") + ylab("score") + ggtitle("Fits score comparison")

vaf_plot <-
  ggplot(data = snps_het, aes(x = as.numeric(DV) / as.numeric(DP))) +
  geom_histogram(binwidth =  0.01, alpha = 0.8) + theme_bw() + xlab("VAF") + ggtitle("Germline SNPs VAF plot")
  
score_plot
vaf_plot
```


```{r}

cna <-  example$cna

```


```{r}
tarone_plot <- Tarone_Boostrap(bmix_data_s %>% mutate(NV = DV, VAF = NV / DP))

cna_p <-  cna %>% filter(!is.na(Major)) %>% mutate(karyotype = paste0(Major, ":", minor), length = to - from) %>%
  mutate(is_diploid = ifelse(karyotype == "1:1", "diploid", "aneuploid"), length = to - from ) %>%
  dplyr::select(length, is_diploid) %>% group_by(is_diploid) %>%  summarize(t_length = sum(length)) %>%
  mutate(t_length =(t_length / sum(t_length)) %>% round(2))


blank_theme <- theme_minimal()+
  theme(
    axis.title.x = element_blank(),
    axis.title.y = element_blank(),
    panel.border = element_blank(),
    panel.grid=element_blank(),
    axis.ticks = element_blank(),
    plot.title=element_text(size=14, face="bold")
  )

cna_p$position <-  c(0,cumsum(cna_p$t_length)[1]) + cna_p$t_length/2

cna_p$is_diploid <-  factor(cna_p$is_diploid, levels = c("diploid", "aneuploid"))

cna_plot <-  ggplot(cna_p, aes(x="", y=t_length, fill=is_diploid) ) +
  geom_bar(width = 1, stat = "identity") + coord_polar("y", direction = 1) +
   blank_theme +
  theme(axis.text.x=element_blank()) +
  geom_text(aes(y = position,
                label = scales::percent(t_length)), size=4) + ggtitle("Percentage of diploid genome") +
  scale_fill_manual(values = c("orange", "gainsboro"))

cna_plot
```

